src/compiler/programDiagnostics.ts

```diff
@@ -0,0 +1,425 @@
+                                diagnostic.fileProcessingReason,
+                                diagnostic.diagnostic,
+                                diagnostic.args || emptyArray,
+                            ),
+                        );
+                    case FilePreprocessingDiagnosticsKind.FilePreprocessingLibReferenceDiagnostic:
+                        return computedDiagnostics!.add(filePreprocessingLibreferenceDiagnostic(program, diagnostic));
+                    case FilePreprocessingDiagnosticsKind.ResolutionDiagnostics:
+                        return diagnostic.diagnostics.forEach(d => computedDiagnostics!.add(d));
+                    default:
+                        Debug.assertNever(diagnostic);
+                }
+            });
+            lazyConfigDiagnostics?.forEach(({ file, diagnostic, args }) =>
+                computedDiagnostics!.add(
+                    createDiagnosticExplainingFile(program, file, /*fileProcessingReason*/ undefined, diagnostic, args),
+                )
+            );
+            fileReasonsToChain = undefined;
+            reasonToRelatedInfo = undefined;
+            return computedDiagnostics;
+        },
+    };
+
+    function filePreprocessingLibreferenceDiagnostic(program: Program, { reason }: FilePreprocessingLibReferenceDiagnostic) {
+        const { file, pos, end } = getReferencedFileLocation(program, reason) as ReferenceFileLocation;
+        const libReference = file.libReferenceDirectives[reason.index];
+        const libName = getLibNameFromLibReference(libReference);
+        const unqualifiedLibName = removeSuffix(removePrefix(libName, "lib."), ".d.ts");
+        const suggestion = getSpellingSuggestion(unqualifiedLibName, libs, identity);
+        return createFileDiagnostic(
+            file,
+            Debug.checkDefined(pos),
+            Debug.checkDefined(end) - pos,
+            suggestion ? Diagnostics.Cannot_find_lib_definition_for_0_Did_you_mean_1 : Diagnostics.Cannot_find_lib_definition_for_0,
+            libName,
+            suggestion!,
+        );
+    }
+
+    function createDiagnosticExplainingFile(program: Program, file: SourceFile | undefined, fileProcessingReason: FileIncludeReason | undefined, diagnostic: DiagnosticMessage, args: DiagnosticArguments): Diagnostic {
+        let seenReasons: Set<FileIncludeReason> | undefined;
+        const reasons = file && fileReasons.get(file.path);
+        let fileIncludeReasons: DiagnosticMessageChain[] | undefined;
+        let relatedInfo: DiagnosticWithLocation[] | undefined;
+        let locationReason = isReferencedFile(fileProcessingReason) ? fileProcessingReason : undefined;
+        let fileIncludeReasonDetails: DiagnosticMessageChain | undefined;
+        let redirectInfo: DiagnosticMessageChain[] | undefined;
+        let cachedChain = file && fileReasonsToChain?.get(file.path);
+        let chain: DiagnosticMessageChain | undefined;

```

Comment:
It's kind of hard to see where actual code gets run, versus uninitialized state.

```suggestion
        let seenReasons: Set<FileIncludeReason> | undefined;
        let fileIncludeReasons: DiagnosticMessageChain[] | undefined;
        let relatedInfo: DiagnosticWithLocation[] | undefined;
        let fileIncludeReasonDetails: DiagnosticMessageChain | undefined;
        let redirectInfo: DiagnosticMessageChain[] | undefined;
        let chain: DiagnosticMessageChain | undefined;
        
        const reasons = file && fileReasons.get(file.path);
        let locationReason = isReferencedFile(fileProcessingReason) ? fileProcessingReason : undefined;
        let cachedChain = file && fileReasonsToChain?.get(file.path);
```